### PR TITLE
Lazy loading tests

### PR DESCRIPTION
**Pull request by hendrikebbers**
ClientSessionExecutor support in controller tests

---
This change is [Reviewable](https://reviewable.io/reviews/canoo/dolphin-platform/755)

### CODE REVIEW

File: platform-integration-tests/integration-server/src/main/java/com/canoo/dolphin/integration/server/runlater/RunLaterTestController.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static com.canoo.dolphin.integration.runlater.RunLaterTestConstants.RUN_LATER_ACTION_NAME;
+import static com.canoo.dolphin.integration.runlater.RunLaterTestConstants.RUN_LATER_ASYNC_ACTION_NAME;
+import static com.canoo.dolphin.integration.runlater.RunLaterTestConstants.RUN_LATER_CONTROLLER_NAME;
+
+@DolphinController(RUN_LATER_CONTROLLER_NAME)
+public class RunLaterTestController {
+
+    private AtomicInteger callIndex = new AtomicInteger();
```
Comment (hendrikebbers) [inline platform-integration-tests/integration-server/src/main/java/com/canoo/dolphin/integration/server/runlater/RunLaterTestController.java]: final

File: platform/dolphin-platform-spring-controller-tester/src/main/java/com/canoo/impl/dp/spring/test/ClientTestFactory.java [hunk 2]
```diff
@@ -48,6 +52,47 @@ public void invoke(String actionName, Param... params) {
                     }
                 }
 
+                @Override
+                public AsyncCondition createAsyncCondition() {
+                    return new AsyncCondition() {
```
Comment (hendrikebbers) [inline platform/dolphin-platform-spring-controller-tester/src/main/java/com/canoo/impl/dp/spring/test/ClientTestFactory.java]: AsyncConditionImpl class

File: platform/dolphin-platform-spring-controller-tester/src/main/java/com/canoo/impl/dp/spring/test/ClientTestFactory.java [hunk 3]
```diff
@@ -48,6 +52,47 @@ public void invoke(String actionName, Param... params) {
...
+                            while (ping.get()) {
+                                sendPing();
+
+                                Thread.sleep(100);
+                            }
+                        }
+
+                        @Override
+                        public void await(long time, TimeUnit unit) throws InterruptedException, TimeoutException {
+                            long endTime = System.currentTimeMillis() + unit.toMillis(time);
```
Comment (hendrikebbers) [inline platform/dolphin-platform-spring-controller-tester/src/main/java/com/canoo/impl/dp/spring/test/ClientTestFactory.java]: final

File: platform/dolphin-platform-spring-controller-tester/src/main/java/com/canoo/impl/dp/spring/test/ClientTestFactory.java [hunk 4]
```diff
@@ -48,6 +52,47 @@ public void invoke(String actionName, Param... params) {
...
+                            ping.set(true);
+                            while (ping.get()) {
+                                sendPing();
+
+                                Thread.sleep(100);
+                            }
+                        }
+
+                        @Override
+                        public void await(long time, TimeUnit unit) throws InterruptedException, TimeoutException {
```
Comment (hendrikebbers) [inline platform/dolphin-platform-spring-controller-tester/src/main/java/com/canoo/impl/dp/spring/test/ClientTestFactory.java]: Assert.notNull

File: platform/dolphin-platform-spring-controller-tester/src/main/java/com/canoo/impl/dp/spring/test/TestClientContextImpl.java [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+import com.canoo.dp.impl.client.legacy.communication.AbstractClientConnector;
+import com.canoo.dp.impl.remoting.legacy.util.Function;
+import com.canoo.platform.client.ClientConfiguration;
+import com.canoo.platform.client.session.ClientSessionStore;
+
+import java.net.URL;
+
+public class TestClientContextImpl extends ClientContextImpl implements TestClientContext {
+
+    public TestClientContextImpl(ClientConfiguration clientConfiguration, URL endpoint, Function<ClientModelStore, AbstractClientConnector> connectorProvider, ClientSessionStore clientSessionStore) {
```
Comment (hendrikebbers) [inline platform/dolphin-platform-spring-controller-tester/src/main/java/com/canoo/impl/dp/spring/test/TestClientContextImpl.java]: final

File: platform/dolphin-platform-spring-controller-tester/src/main/java/com/canoo/impl/dp/spring/test/TestDolphinContext.java [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+import java.util.concurrent.LinkedBlockingQueue;
+
+/**
+ * Created by hendrikebbers on 05.12.17.
+ */
+public class TestDolphinContext extends DolphinContext {
+
+    private BlockingQueue<Runnable> callLaterTasks = new LinkedBlockingQueue<>();
+
+    public TestDolphinContext(RemotingConfiguration configuration, ClientSession clientSession, ClientSessionProvider clientSessionProvider, ManagedBeanFactory beanFactory, ControllerRepository controllerRepository, Callback<DolphinContext> onDestroyCallback) {
```
Comment (hendrikebbers) [inline platform/dolphin-platform-spring-controller-tester/src/main/java/com/canoo/impl/dp/spring/test/TestDolphinContext.java]: final

File: platform/dolphin-platform-spring-controller-tester/src/main/java/com/canoo/impl/dp/spring/test/TestDolphinContext.java [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Future;
+import java.util.concurrent.LinkedBlockingQueue;
+
+/**
+ * Created by hendrikebbers on 05.12.17.
+ */
+public class TestDolphinContext extends DolphinContext {
+
+    private BlockingQueue<Runnable> callLaterTasks = new LinkedBlockingQueue<>();
```
Comment (hendrikebbers) [inline platform/dolphin-platform-spring-controller-tester/src/main/java/com/canoo/impl/dp/spring/test/TestDolphinContext.java]: final

File: platform/dolphin-platform-spring-controller-tester/src/main/java/com/canoo/impl/dp/spring/test/TestDolphinContext.java [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+        getServerConnector().getRegistry().register(PingCommand.class, new CommandHandler() {
+            @Override
+            public void handleCommand(Command command, List response) {
+
+            }
+        });
+    }
+
+    @Override
+    public List<Command> handle(List<Command> commands) {
```
Comment (hendrikebbers) [inline platform/dolphin-platform-spring-controller-tester/src/main/java/com/canoo/impl/dp/spring/test/TestDolphinContext.java]: final

**Comment by hendrikebbers** [comment]
fix for https://github.com/canoo/dolphin-platform/issues/751

File: platform-integration-tests/integration-server/src/test/java/com/canoo/dolphin/integration/server/runLater/RunLaterControllerTest.java [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+        Assert.assertTrue(model.getActionPreRunLaterCallIndex() < model.getActionPostRunLaterCallIndex());
+        Assert.assertTrue(model.getActionPostRunLaterCallIndex() < model.getActionRunLaterCallIndex());
+    }
+
+    @Test
+    public void testRunLaterAsyncInAction() throws Exception {
+        //given:
+        final ControllerUnderTest<RunLaterTestBean> controller = createController(RUN_LATER_CONTROLLER_NAME);
+        final RunLaterTestBean model = controller.getModel();
+        final AsyncCondition condition = controller.createAsyncCondition();
```
Comment (hendrikebbers) [inline platform-integration-tests/integration-server/src/test/java/com/canoo/dolphin/integration/server/runLater/RunLaterControllerTest.java]: Rename AsyncCondition to ControllerMonitor

File: platform-integration-tests/integration-server/src/main/java/com/canoo/dolphin/integration/server/runlater/RunLaterTestController.java (hunk 1 above)
Comment (hendrikebbers) [reply platform-integration-tests/integration-server/src/main/java/com/canoo/dolphin/integration/server/runlater/RunLaterTestController.java]: done

File: platform/dolphin-platform-spring-controller-tester/src/main/java/com/canoo/impl/dp/spring/test/ClientTestFactory.java (hunk 2 above)
Comment (hendrikebbers) [reply platform/dolphin-platform-spring-controller-tester/src/main/java/com/canoo/impl/dp/spring/test/ClientTestFactory.java]: done

File: platform/dolphin-platform-spring-controller-tester/src/main/java/com/canoo/impl/dp/spring/test/ClientTestFactory.java (hunk 4 above)
Comment (hendrikebbers) [reply platform/dolphin-platform-spring-controller-tester/src/main/java/com/canoo/impl/dp/spring/test/ClientTestFactory.java]: done

File: platform/dolphin-platform-spring-controller-tester/src/main/java/com/canoo/impl/dp/spring/test/ClientTestFactory.java (hunk 3 above)
Comment (hendrikebbers) [reply platform/dolphin-platform-spring-controller-tester/src/main/java/com/canoo/impl/dp/spring/test/ClientTestFactory.java]: done

File: platform/dolphin-platform-spring-controller-tester/src/main/java/com/canoo/impl/dp/spring/test/TestClientContextImpl.java (hunk 5 above)
Comment (hendrikebbers) [reply platform/dolphin-platform-spring-controller-tester/src/main/java/com/canoo/impl/dp/spring/test/TestClientContextImpl.java]: done

File: platform/dolphin-platform-spring-controller-tester/src/main/java/com/canoo/impl/dp/spring/test/TestDolphinContext.java (hunk 7 above)
Comment (hendrikebbers) [reply platform/dolphin-platform-spring-controller-tester/src/main/java/com/canoo/impl/dp/spring/test/TestDolphinContext.java]: done

File: platform/dolphin-platform-spring-controller-tester/src/main/java/com/canoo/impl/dp/spring/test/TestDolphinContext.java (hunk 6 above)
Comment (hendrikebbers) [reply platform/dolphin-platform-spring-controller-tester/src/main/java/com/canoo/impl/dp/spring/test/TestDolphinContext.java]: done

File: platform/dolphin-platform-spring-controller-tester/src/main/java/com/canoo/impl/dp/spring/test/TestDolphinContext.java (hunk 8 above)
Comment (hendrikebbers) [reply platform/dolphin-platform-spring-controller-tester/src/main/java/com/canoo/impl/dp/spring/test/TestDolphinContext.java]: done

File: platform/dolphin-platform-spring-controller-tester/src/main/java/com/canoo/impl/dp/spring/test/ClientTestFactory.java [hunk 10]
```diff
@@ -48,6 +52,47 @@ public void invoke(String actionName, Param... params) {
...
+                            long endTime = System.currentTimeMillis() + unit.toMillis(time);
+                            ping.set(true);
+                            while (ping.get() && System.currentTimeMillis() < endTime) {
+                                sendPing();
```
[the remark this answers is not in the document]
Comment (hendrikebbers) [reply platform/dolphin-platform-spring-controller-tester/src/main/java/com/canoo/impl/dp/spring/test/ClientTestFactory.java]: Done

**Comment by blackdrag** [comment]
Review status: 0 of 22 files reviewed at latest revision, 11 unresolved discussions, some commit checks broke.

---

*[platform-integration-tests/integration-server/src/test/java/com/canoo/dolphin/integration/server/runLater/RunLaterControllerTest.java, line 68 at r2](https://reviewable.io:443/reviews/canoo/dolphin-platform/755#-L-kujtC-btd13V8iy0z:-L-kujtC-btd13V8iy1-:b-aq1c8u) ([raw file](https://github.com/canoo/dolphin-platform/blob/bcaa5387d8b164aef6191a9f297d065648cac4b9/platform-integration-tests/integration-server/src/test/java/com/canoo/dolphin/integration/server/runLater/RunLaterControllerTest.java#L68)):*
> ```Java
>         final ControllerUnderTest<RunLaterTestBean> controller = createController(RUN_LATER_CONTROLLER_NAME);
>         final RunLaterTestBean model = controller.getModel();
>         final CommunicationMonitor condition = controller.createMonitor();
> ```

you should rename the variable as well

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/canoo/dolphin-platform/755)*

**Comment by blackdrag** [comment]
Reviewed 13 of 22 files at r1, 10 of 10 files at r2.
Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks broke.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/canoo/dolphin-platform/755)*

**Pull request merged**